### ISSUE TITLE
Classification performance.

### ISSUE THREAD

**Issue by memecode** [freiz/terminator#4]
Either I don't understand how to drive this lib. Or it's not operating correctly. Probably more "a" hopefully.

So here is what I see when I train a decent size of spam and ham email from my personal folders:
![graph](https://i.imgur.com/OmCCdwF.png)

This is the sorted scores of all the spam and inbox email that I trained the filter on. The DB file is 23MB which sounds plausible. There doesn't seem to be a significant difference between the two groups of email. Basically my code parses the email into a list of words, removing all the headers and HTML encoding. Then it passes that as the content std::string to Terminator::Train with the spam bool set. Then I went back and called Predict on the same email to make the graphs in the image.

I did notice that the classifier_weights_ seem to be weird:
classifier_weights_[0]=-1.25549e+65
classifier_weights_[1]=0
classifier_weights_[2]=0
classifier_weights_[3]=0
classifier_weights_[4]=0
classifier_weights_[5]=0
classifier_weights_[6]=0
classifier_weights_[7]=0

That looks wrong... but it's what gets saved and reloaded... maybe that's part of my issue?

**Comment by memecode** [comment]
Interestingly in the demo code you seem to pass the raw rfc822 email image, headers and all into the training and classification methods. I should try that and see what difference it makes...!

**Comment by memecode** [comment]
So I've tried just the rfc822 format directly and normalized the axis of the graph to 0:1:

![graph](https://i.imgur.com/9Y5CvAr.png)

It's not different enough to warrant closing the issue though...

**Comment by memecode** [comment]
Btw the demo app got these results:

Average: 9.6 (e-mails per second)
ham->ham  =  4134               spam->spam =  1731
ham->spam =   15                spam->ham  =  154

**Comment by freiz** [comment]
Hello, thanks for playing with it. 

It is a project during my master over 10 years ago, do not remember every detail but I will try my best to provide help. The classification performance is evaluated over all the public datasets as well as some private datasets (e.g. SEWM competition and private corpus from Netease), so do not need to worry.

From the result of your method, it is highly possible a misuse, typically a trained case will be perfectly predicted (it is generally not a good idea to evaluate over trained cases), also with the sorted values I cannot see much interesting info.

For a better experiment setting, you can do either the below.
1. Simulate online scenarios with full feedback (like the demo case). For each email, first, call predict and then call train with the correct label.
2. Shuffle your dataset and split it to 9:1 (training/validation), call train for each training case, and predict on the separate validation set.
Then evaluate by ROCA, precision/recall, F-score, etc.

Note, the library only accepts the raw email content (BASE64 encoded, does not require any preprocessing). let me know if it works for you, or you can share with me a sample code I will take a look.

**Comment by memecode** [comment]
I got my email client to write out the folders to the same format as the demo uses (index+rfc822 fmt in subfolders) and ran the demo over that:

ham->ham  =  7380               spam->spam =  1188
ham->spam =    7                spam->ham  =  352

The spam->ham number is pretty bad. 23% error rate? Is that the correct way of reading it?

**Comment by freiz** [comment]
From my point of view, it is a pretty good result. The demo code is standard full feedback online learning task without prior knowledge for benchmarking purposes, i.e. it starts prediction from nothing (I have not provided a pre-trained model here). In a real-world scenario, you will first train it before deploying it. 

You can try training with the spam assassin dataset before the online predicting/training. 
Or like I suggested, split your dataset into the training set and the validation set. The result will be much better since avoiding the cold-start issue.
You can also try to update the threshold for a better understanding of training/predicting error, the default threshold I use is over 0.5 because the cost is much larger if you make mistake on a ham email compared to a spam email.
